gpuweb/cts
# Parallelize Grunt build

**Commits may be reviewed separately**

Parallelizing speeds up build time by about 3x on my machine (106s -> 36s). The parallelization here is very ad-hoc (compared to a real build system), but it should be reasonably fast. Memory usage peaks at about 2.0GB on my system (though I have also seen larger numbers, up to 3.8GB - unsure if it was caused by this build).

I also tried running eslint as a separate step at the end, because it uses a lot of memory. This still gets a speedup of 2x (106s -> 54s) on my machine with a somewhat lower peak memory usage. eslint also appears to be the longest pole in the build. Future work, then, to try to speed up eslint: #3135

I had hoped this would alleviate some pain from requiring PRs to be up-to-date before merging, but unfortunately the GitHub Actions VMs seem to be resource limited in a such a way that this doesn't really help. Still, it's very nice locally.

Also does some cleanup of the Gruntfile that was helpful to do first:
- Reduce ordering hazards between Grunt tasks by not generating files on top of existing files
- Grunt task changes
  - Remove partial checks from `grunt standalone` and `grunt wpt`
  - Make `grunt unittest` only run unittests
  - Rename `grunt check` to `grunt typecheck`
  - Add `grunt checks` to run all checks`
  - Rename `grunt pre` to `grunt all` and make `grunt pre` an alias
  - Improve output of `grunt default` (aka `grunt`)
- Add a little documentation for the TSDoc docs

Issue: None

<hr>

**Requirements for PR author:**

- [x] All missing test coverage is tracked with "TODO" or `.unimplemented()`.
- [x] New helpers are `/** documented */` and new helper files are found in `helper_index.txt`.
- [x] Test behaves as expected in a WebGPU implementation. (If not passing, explain above.)

**Requirements for [reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md):**

- [ ] Tests are properly located in the test tree.
- [ ] [Test descriptions](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) allow a reader to "read only the test plans and evaluate coverage completeness", and accurately reflect the test code.
- [ ] Tests provide complete coverage (including validation control cases). **Missing coverage MUST be covered by TODOs.**
- [ ] Helpers and types promote readability and maintainability.

When landing this PR, be sure to make any necessary issue status updates.
